### PR TITLE
HL-312, HL-315

### PR DESCRIPTION
**Pull request by bitcoinbrisbane**
- Out by 1 errors on NFT minting
- Settle all function HL-315
- Add time out function HL-312
- Reduce logic
- Fixes to unit tests
- Make unit tests name casing consistent
- Fix unit test gas pricing

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -292,59 +287,84 @@ contract Market is IMarket, Ownable, ERC721 {
...
 
-		_burn(id);
+		IERC20(underlying).transfer(recipient, _bets[index].payout);
+		_burn(index);
 
-		emit Settled(id, _bets[id].payout, result, _bets[id].owner);
+		emit Settled(index, _bets[index].payout, result, recipient);
+	}
+
+	function settleMarket(bytes16 marketId) external {
```
Comment (bazmatic) [inline contracts/Market.sol:356]: Should we add this to `IMarket` ?

**Review by bazmatic — DISMISSED** [review]
Should we add `settleMarket()` to `IMarket`?

File: contracts/Market.sol [hunk 2]
```diff
@@ -247,27 +247,22 @@ contract Market is IMarket, Ownable, ERC721 {
...
 		return _back(
-			nonce,
 			propositionId,
 			marketId,
 			wager,
-			odds,
 			close,
 			end,
-			signature
+			payout
```
Comment (brettvniekerk) [inline contracts/Market.sol:255]: backing no longer requires a signature?

File: contracts/Market.sol [hunk 3]
```diff
@@ -292,59 +287,84 @@ contract Market is IMarket, Ownable, ERC721 {
 		// add to the total potential payout for this proposition
 		_potentialPayout[propositionId] += payout;
 
+		uint64 index = _getCount();
```
Comment (brettvniekerk) [inline contracts/Market.sol:290]: the `Placed` event previously emitted `_getCount() - 1`, just wondering why we now want to emit `_getCount()`

File: contracts/Market.sol [hunk 4]
```diff
@@ -292,59 +287,84 @@ contract Market is IMarket, Ownable, ERC721 {
...
+
 		_bets.push(
 			Bet(propositionId, marketId, wager, payout, end, false, _msgSender())
 		);
 
-		// use _getCount() to avoid stack too deep
-		_marketBets[marketId].push(_getCount());
-		_mint(_msgSender(), _getCount() - 1);
+		_marketBets[marketId].push(index);
+		_mint(_msgSender(), index);
```
Comment (brettvniekerk) [inline contracts/Market.sol:297]: why not `msg.sender`?

File: contracts/Market.sol (hunk 1 above)
Comment (brettvniekerk) [reply contracts/Market.sol:356]: +1

File: contracts/Market.sol [hunk 5]
```diff
@@ -292,59 +287,84 @@ contract Market is IMarket, Ownable, ERC721 {
...
 
-		emit Settled(id, _bets[id].payout, result, _bets[id].owner);
+		emit Settled(index, _bets[index].payout, result, recipient);
+	}
+
+	function settleMarket(bytes16 marketId) external {
+		uint64[] memory bets = _marketBets[marketId];
+		uint256 total = bets.length;
+
+		for (uint64 i = 0; i < total; i++) {
```
Comment (brettvniekerk) [inline contracts/Market.sol:360]: won't this function run out of gas as the amount of bets increases?

Comment (brettvniekerk) [reply contracts/Market.sol:360]: nvm

File: contracts/Market.sol (hunk 3 above)
Comment (Full-Node-Capital) [reply contracts/Market.sol:290]: I believe the index was out by 1

File: contracts/MarketWithRisk.sol [hunk 6]
```diff
@@ -39,18 +49,29 @@ contract MarketWithRisk is Market {
...
 			close,
 			end,
-			signature
+			risk
+		));
+
+		require(isValidSignature(messageHash, signature) == true, "back: Invalid signature");
+
+		uint256 payout = wager * _getOddsWithRisk(wager, odds, propositionId, marketId, risk);
+		return _back(
```
Comment (brettvniekerk) [inline contracts/MarketWithRisk.sol:68]: could we just override the `back` function in this contract to use risk?

File: contracts/Market.sol (hunk 2 above)
Comment (bazmatic) [reply contracts/Market.sol:255]: That's just the internal `_back()` method

**Review by brettvniekerk — DISMISSED** [review]
small comments otherwise LGTM

File: contracts/Market.sol (hunk 4 above)
Comment (Full-Node-Capital) [reply contracts/Market.sol:297]: To stop sneakyness https://forum.openzeppelin.com/t/why-use-msgsender-rather-than-msg-sender/4370

**Review by Full-Node-Capital — DISMISSED** [review]
Besides adding that interface, LGTM

File: contracts/Market.sol (hunk 2 above)
Comment (brettvniekerk) [reply contracts/Market.sol:255]: ah so the signature is checked in the public `back` function then?

File: contracts/Market.sol (hunk 1 above)
Comment (bitcoinbrisbane) [reply contracts/Market.sol:356]: Done

File: contracts/MarketWithRisk.sol (hunk 6 above)
Comment (bitcoinbrisbane) [reply contracts/MarketWithRisk.sol:68]: No because we have an extra param `risk`

**Review by bazmatic — APPROVED** [review]

**Review by Full-Node-Capital — DISMISSED** [review]

File: scripts/settle.py [hunk 7]
```diff
@@ -116,7 +116,7 @@ def update_market_oracle(market_address, oracle):
             bet = market.functions.getBetByIndex(i).call()
 
             # check if bet is less than 2 hours old
-            if bet[2] > now - 60 * 60 * 24 and bet[3] == False:
+            if bet[2] > now - 60 * 60 * 72 and bet[3] == False:
```
Comment (brettvniekerk) [inline scripts/settle.py:119]: does this check if a bet is less than 2 hours old?

Comment (bazmatic) [reply scripts/settle.py:119]: Looks like 3 days...

Comment (bitcoinbrisbane) [reply scripts/settle.py:119]: Should have pushed that.  Fixed

**Review by brettvniekerk — DISMISSED** [review]

**Review by OptimusOpus — APPROVED** [review]

**Pull request merged**